File: algo_library/python/graph/csr.py
# 無向グラフの場合は graphEdges にいつも通り u -> v と v -> u の両方を加える
# 重み付きグラフの場合は endList に格納する値を (終点, 重み) にすればよい

# 有向グラフのとき

# N := 頂点数
# graphEdges[i] := [辺 i の始点, 辺 i の終点]
def csr(N: int, graphEdges: list[list[int]]):
  start = [0]*(N+1)
  # endList := 辺を始点の昇順にソートした時の終点のリスト
  endList = [0]*len(graphEdges)

  # start[i+1] := 頂点 i を始点とする辺の数 となるように構築
  for u,v in graphEdges: # u -> v への辺
    start[u+1] += 1
  # start[i] := 始点の頂点番号が i 未満である辺の数 になるように累積和を取る
  # このとき、start[i] = endList で頂点 i が始点となる最小の index (0-index)
  for i in range(1,N+1):
    start[i] += start[i-1]

  # pointer[i] := 始点を i とする辺の endList への挿入位置
  pointer = start[:]
  for u,v in graphEdges:
    endList[pointer[u]] = v
    pointer[u] += 1
    
  return start, endList

File: algorithm_library/cpp-from-py/graph/CSR.cpp
#include <bits/stdc++.h>
using namespace std;

pair<vector<int>, vector<int>> csr(int N, const vector<pair<int, int>>& graphEdges) {
    vector<int> start(N + 1, 0), endList(graphEdges.size());
    for (auto [u, v] : graphEdges) start[u + 1]++;
    for (int i = 1; i <= N; i++) start[i] += start[i - 1];
    vector<int> pointer = start;
    for (auto [u, v] : graphEdges) endList[pointer[u]++] = v;
    return {start, endList};
}

template <class Cost = long long>
struct Graph {
    int N, M;
    vector<int> start;
    vector<tuple<int, int, optional<Cost>>> buf;
    vector<pair<int, optional<Cost>>> endList;

    Graph(int n, int m) : N(n + 1), M(m), start(N + 1, 0), endList(m) {}

    void add_edge(int u, int v, optional<Cost> cost = nullopt) {
        buf.emplace_back(u, v, cost);
        start[u]++;
        if ((int)buf.size() == M) build();
    }

    void build() {
        partial_sum(start.begin(), start.end(), start.begin());
        for (auto [u, v, c] : buf) {
            start[u]--;
            endList[start[u]] = {v, c};
        }
    }

    vector<pair<int, optional<Cost>>> get_edges(int u) const {
        return vector<pair<int, optional<Cost>>>(endList.begin() + start[u], endList.begin() + start[u + 1]);
    }
};
